### DISCUSSION TITLE
Loaded GPX Route not obeyed

### DISCUSSION THREAD

**Discussion by MotoUKRider** [devemux86/cruiser#39] [Q&A] [answered]
When loading a GPX from Garmin's BaseCamp, Cruiser app does not produce the correct route, and even deletes some waypoint/stops from the route.  The overlay appears OK, showing where the route *should* have gone, but there is something wrong somewhere.

I notice that the overlay/route dialog has been simplified, but previous build had no issue.  I believe that I have got my settings correct as the new advanced-level settings layout changed a few things for me. Settings seems much nicer, btw :)

Old v2 build which I'm subbed to loads the route fine.

Without finding an APK extractor, how does one roll-back to a more workable build?!  This is one of the problems I have with GooglePlay - if an app update is buggy/poor, then the user is (to quote a technical term) "stuffed".

**Comment by MotoUKRider** [comment]
Here's an example - my Scotland route completely misses that isolated Favorite near the coast, so I've drawn the actual BaseCamp/GPX route to show what it missed.  I understand that a routing engine may force you onto a faster road when one exists (adding more shaping points is understandable), but to miss explicit instructions like this?  Something screwy :(

![missed-waypoint](https://user-images.githubusercontent.com/125696109/219780865-933ef599-40dd-458e-8447-85e40f6ffb6e.jpg)

**Comment by devemux86** [comment]
@MotoUKRider thanks for the report!

Do not worry, nothing is final and everything can change again!

The export / import dialogs were "simplified" because the previous implementation was too complicated
for many users to understand how to use and also to maintain the multiple options efficiently in the code.

So we better design our options carefully based on your real requirements. :slightly_smiling_face: 

Is it the previous option to choose the maximum number of waypoints that you are missing?
Algorithm now selects it automatically based on track distance, approximately every ~10km.

Should we adjust the algorithm or return the max waypoints option?

Also do you have a sample GPX file to test?

**Comment by MotoUKRider** [comment]
The Waypoint number slider wasn't something I used - but whatever the new algorithm is doing, it's certainly not quite right :)  This is with BRouter (my default) but also GraphHopper produces same issue.  A short local route with 16 routing-points (5 of which are Waypoints) from BaseCamp also produced a missed Waypoint and deleted/ignored shaping-points.

Sample GPX in a ZIP file showing 1st part of a Scotland trip which is from the screenshot above:
[gpx-sample.zip](https://github.com/devemux86/cruiser/files/10771504/gpx-sample.zip)

**Reply by devemux86** [reply]
Thanks for the sample.

It does not have a track, it seems to have 7 (wpt) waypoints and 27 (rtept) route points.
And you want to keep all 27 route points in the route calculation?

**Comment by MotoUKRider** [comment]
Yes please - it's what comes out of BaseCamp as the default GPX (so the current state of Cruiser ruins things for anyone else also having the same workflow).  It's shaping-point heavy because fun, twisty roads that I've researched and Via Points to stop at.  This would load just fine in the previous build.  This kind of route could also be build very quickly using move+tap within your app - it works perfectly in that regard :)

My route, for example, has a few places where a simple straight line would get somewhere much quicker, but the additional shaping-points help to follow the coast or keep on a much more adventurous road. The removal of a defined Via/Waypoint is a concern.  Of course the end result may vary if I were to give the file to a Calimoto, Rever, CoPilot, TomTom or Waze user.  I find that placing a point just after a turn works best.  There have been times where I've had to place extra shaping points because Cruiser has found a faster way, but I didn't want to travel on it.

As long as it faithfully reproduces the GPX file, doesn't adios any of my Waypoints and calculates the best way between each routing+waypoint, I'll be happy :)

**Reply by devemux86** [reply]
Thanks for the explanation, I understand what you mean.

I will check again and improve the procedure
and a new version can always be released soon!

**Reply by MotoUKRider** [reply]
Excellent - here's another with the route being "creatively recalculated"!  It deletes the "Oak and Ivy" pub from the route, and even ignores a set of shaping-points in+out of Winchelsea by taking me on the *same* road in+out to/from Point 1.  Very odd.

This reminds me of riding rule 3847: Never update your navigation app the day before a tour ;)

[gpx-sample-2.zip](https://github.com/devemux86/cruiser/files/10771817/gpx-sample-2.zip)

**Reply by devemux86** [reply]
Thanks, this is also an example with (rtept) route points.
I will fix the route points to not skip them during import.

@MotoUKRider 
Do you have a sample with (trk) track that you see a difference in new Cruiser or tracks work properly?

**Reply by MotoUKRider** [reply]
No, sorry. I don't use tracks as they serve no real purpose for me.

**Comment by devemux86** [comment]
@MotoUKRider 

Can you try GPX route import with desktop [Cruiser 3.0.10 (Beta)](https://www.dropbox.com/sh/914t158im4b7lxr/AADOeSE1UjaW4FS2PLc7kwAma)?

**Reply by MotoUKRider** [reply]
It's fine with those 2 examples and others in my collection - GPX wasn't an issue I could observe in desktop .8+9 because of the OpenGL startup crash - I never got as far as importing a route :(  However, the program loads just fine, routes perfectly on BRouter+OpenRoute online and that multi-item popup box crash is gone... Happy days!

**Reply by devemux86** [reply]
Thanks for the report and the feedback!

The fix will also be included in Android Cruiser 3.0.10 (coming soon).

**Reply by airald** [reply]
I hope it be released soon. The import menu options are the way to import waypoints with name instead of "waypoint xxx". I use it a lot, when I plan a route, I have anotations of things I can visit during route, places, etc... I use Garmin format made by Furkot planner.

**Reply by devemux86** [reply]
@airald 

Can you post a GPX sample to test and understand better?
What is the content of the GPX file: waypoints or route points or track or a combination?
And which of these do you want to use for routing?

**Reply by airald** [reply]
Hello, 

I place waypoints and I use it to make the route. The file contains waypoints and also the track. I use track+waypoints to show as overlay and route for routing
![Screenshot_20230219_203316_Cruiser](https://user-images.githubusercontent.com/20151719/219973086-daadc671-47e3-4bf1-9e09-b38e69e2c64f.jpg)
this is the result
![Screenshot_20230219_210216_Cruiser](https://user-images.githubusercontent.com/20151719/219973099-7ad3be8a-e7eb-43f5-8bfa-6bb1b75c10b3.jpg)
and waypoints have the original name
![Screenshot_20230219_211142_Cruiser](https://user-images.githubusercontent.com/20151719/219973111-98ed0316-d81e-413b-9e21-572e7ca00d97.jpg)
(version 3.0.8)

When I do the same using version 3.0.9
![Screenshot_2023-02-19-21-08-18-748_gr talent cruiser](https://user-images.githubusercontent.com/20151719/219973134-c1a4f931-db34-4fab-a7ee-28e7352d5a56.jpg)
the route is not the desired and original waypoints are lost.
![Screenshot_2023-02-19-21-12-20-170_gr talent cruiser](https://user-images.githubusercontent.com/20151719/219973177-a2a84410-f3a1-4a45-b561-c1f67fd49d04.jpg)

Hope it can be useful.


[waypoints no name.zip](https://github.com/devemux86/cruiser/files/10777742/waypoints.no.name.zip)

**Reply by devemux86** [reply]
@airald 

Thanks very much for the details and the sample!

I understand what happens, it will work properly again in the next Cruiser update.

**Comment by WolfgangZ42** [comment]
**Sussex Stuff v1.gpx** comes out of BaseCamp and has this special gpx slang with imbedded track informations via <gpxx:rpt lat="51.045076847076416" lon="0.414454936981201" />. Some gpx software doesn't work with this intermix of route and track. For instance kurviger.de doesn't "see" the track in this file. I'm not using BaseCamp at all and therefore take the example file from @MotoUKRider .

To separate route and track _gpxshaping.de_ could be used - source file: **Sussex Stuff v1.gpx** / output file: **Sussex Stuff v1SHP.gpx**, the latter contains some inserted shaping points - not really necessary in this case. Importing **Sussex Stuff v1SHP.gpx** into Cruiser shows a valid route following exactly the track - but all waypoints are lost, instead you get 4 artificial shaping points. (I guess this will be fixed soon.)

I further import **Sussex Stuff v1SHP.gpx** into _kurviger.de_ to get an exact copy as **Sussex Stuff v1SHP.kurviger** (https://kurv.gr/EBa4U) - and if you import this file into Cruiser you get your route, the track as overlay and all waypoints. Okay, you get some shaping points more and all waypoint names start with "xx.x km" - see my **Workflow.pdf**.  

All files are in this zip: [Download.zip](https://github.com/devemux86/cruiser/files/10776036/Download.zip)

**Reply by devemux86** [reply]
@WolfgangZ42 thanks for the detailed example!

> I guess this will be fixed soon

It should work properly in the next Cruiser update.

**Reply by MotoUKRider** [reply]
Interesting stuff - I accept the file is unnecessarily "bloated" for what may be, say, 20 shaping points and a few via/waypoints. That's the only GPX export option from BaseCamp. The main concern was that it worked *perfectly* before this latest build :)

Not wanting to complicate things further, but when I load a GPX from BaseCamp that had *multiple* routes selected from BaseCamp during export, ITNConv allows me to pick a route (or the waypoint list to load).  So, a single GPX file could be a full 7-day tour with 7 routes,   Attached is an example of a 3-route tour of lower Scotland - 3 routes from BaseCamp selected for export.  The resulting GPX is huuuuuge, but offered for analysis.  Garmin: Because even .001' is a turn as far as we're concerned ;)

[multiple-routes.zip](https://github.com/devemux86/cruiser/files/10776564/multiple-routes.zip)

Screen from ITNConv showing what it can "see":
![gpx-multiple-routes-1](https://user-images.githubusercontent.com/125696109/219941168-f7d7d94a-b3d8-4fcd-a7ef-2ee48ecca643.jpg)
Loaded "Day2" from that list:
![gpx-multiple-routes-2](https://user-images.githubusercontent.com/125696109/219941170-806051f1-da08-4be6-8b14-6dc3f2082220.jpg)

PS: In BaseCamp, you can't have duplicate names of *anything*, so you'll see in the routing list that B7042+B70421 and A719+A7192 are actually the same road but have a number suffix.  I obviously have "A7191" in another route somewhere.  This can be annoying if you want to put a custom text tag on a shaping point that you *might* want to switch to a Via Point, eg: "Turn off here" - so you may find that road names/numbers often have a number at the end if they've come from BaseCamp. Fun!

**Reply by devemux86** [reply]
> The main concern was that it worked perfectly before this latest build :)

Please wait for Cruiser 3.0.10, I'm still working on it. :slightly_smiling_face: 

> Not wanting to complicate things further, but when I load a GPX from BaseCamp that had multiple routes selected from BaseCamp during export, ITNConv allows me to pick a route (or the waypoint list to load).

@MotoUKRider thanks for the sample!

_Please can you create a new topic for this?_

It's an interesting idea, so it will not be lost in this discussion.

**Comment by 0709wiwiwi** [comment]
Nice example Wolfgang.

Notice the file "packaging" (by extensions) or an impressively complex
method to obtain just a quite simple track plus via or shaping points
content.

See alone that impressive file introduction.

*"Comparable with an soldier which is overloaded with very many and all
sorts  of decorations"*

The file start before even arriving at any useful content is already quite
crowded.

<?xml version="1.0" encoding="utf-8"?><gpx creator="Garmin Desktop App"
version="1.1" xsi:schemaLocation="http://www.topografix.com/GPX/1/1
http://www.topografix.com/GPX/1/1/gpx.xsd
http://www.garmin.com/xmlschemas/WaypointExtension/v1
http://www8.garmin.com/xmlschemas/WaypointExtensionv1.xsd
http://www.garmin.com/xmlschemas/TrackPointExtension/v1
http://www.garmin.com/xmlschemas/TrackPointExtensionv1.xsd
http://www.garmin.com/xmlschemas/GpxExtensions/v3
http://www8.garmin.com/xmlschemas/GpxExtensionsv3.xsd
http://www.garmin.com/xmlschemas/ActivityExtension/v1
http://www8.garmin.com/xmlschemas/ActivityExtensionv1.xsd
http://www.garmin.com/xmlschemas/AdventuresExtensions/v1
http://www8.garmin.com/xmlschemas/AdventuresExtensionv1.xsd
http://www.garmin.com/xmlschemas/PressureExtension/v1
http://www.garmin.com/xmlschemas/PressureExtensionv1.xsd
http://www.garmin.com/xmlschemas/TripExtensions/v1
http://www.garmin.com/xmlschemas/TripExtensionsv1.xsd
http://www.garmin.com/xmlschemas/TripMetaDataExtensions/v1
http://www.garmin.com/xmlschemas/TripMetaDataExtensionsv1.xsd
http://www.garmin.com/xmlschemas/ViaPointTransportationModeExtensions/v1
http://www.garmin.com/xmlschemas/ViaPointTransportationModeExtensionsv1.xsd
http://www.garmin.com/xmlschemas/CreationTimeExtension/v1
http://www.garmin.com/xmlschemas/CreationTimeExtensionsv1.xsd
http://www.garmin.com/xmlschemas/AccelerationExtension/v1
http://www.garmin.com/xmlschemas/AccelerationExtensionv1.xsd
http://www.garmin.com/xmlschemas/PowerExtension/v1
http://www.garmin.com/xmlschemas/PowerExtensionv1.xsd
http://www.garmin.com/xmlschemas/VideoExtension/v1
http://www.garmin.com/xmlschemas/VideoExtensionv1.xsd" xmlns="
http://www.topografix.com/GPX/1/1" xmlns:xsi="
http://www.w3.org/2001/XMLSchema-instance" xmlns:wptx1="
http://www.garmin.com/xmlschemas/WaypointExtension/v1" xmlns:gpxtrx="
http://www.garmin.com/xmlschemas/GpxExtensions/v3" xmlns:gpxtpx="
http://www.garmin.com/xmlschemas/TrackPointExtension/v1" xmlns:gpxx="
http://www.garmin.com/xmlschemas/GpxExtensions/v3" xmlns:trp="
http://www.garmin.com/xmlschemas/TripExtensions/v1" xmlns:adv="
http://www.garmin.com/xmlschemas/AdventuresExtensions/v1" xmlns:prs="
http://www.garmin.com/xmlschemas/PressureExtension/v1" xmlns:tmd="
http://www.garmin.com/xmlschemas/TripMetaDataExtensions/v1" xmlns:vptm="
http://www.garmin.com/xmlschemas/ViaPointTransportationModeExtensions/v1"
xmlns:ctx="http://www.garmin.com/xmlschemas/CreationTimeExtension/v1"
xmlns:gpxacc="http://www.garmin.com/xmlschemas/AccelerationExtension/v1"
xmlns:gpxpx="http://www.garmin.com/xmlschemas/PowerExtension/v1"
xmlns:vidx1="http://www.garmin.com/xmlschemas/VideoExtension/v1">

  <metadata>
    <link href="http://www.garmin.com">
      <text>Garmin International</text>
    </link>
    <time>2022-12-14T19:16:07Z</time>
    <bounds maxlat="51.055054664611816" maxlon="0.736620426177979"
minlat="50.882728099822998" minlon="0.414454936981201" />
  </metadata>

Op zo 19 feb. 2023 om 06:39 schreef WolfgangZ42 ***@***.***>:

> *Sussex Stuff v1.gpx* comes out of BaseCamp and has this special gpx
> slang with imbedded track informations via <gpxx:rpt
> lat="51.045076847076416" lon="0.414454936981201" />. Some gpx software
> doesn't work with this intermix of route and track. For instance
> kurviger.de doesn't "see" the track in this file. I'm not using BaseCamp
> at all and therefore take the example file from @MotoUKRider
> <https://github.com/MotoUKRider> .
>
> To separate route and track *gpxshaping.de <http://gpxshaping.de>* could
> be used - source file: *Sussex Stuff v1.gpx* / output file: *Sussex Stuff
> v1SHP.gpx*, the latter contains some inserted shaping points - not really
> necessary in this case. Importing *Sussex Stuff v1SHP.gpx* into Cruiser
> shows a valid route following exactly the track - but all waypoints are
> lost, instead you get 4 artificial shaping points. (I guess this will be
> fixed soon.)
>
> I further import *Sussex Stuff v1SHP.gpx* into *kurviger.de
> <http://kurviger.de>* to get an exact copy as *Sussex Stuff
> v1SHP.kurviger* (https://kurv.gr/EBa4U) - and if you import this file
> into Cruiser you get your route, the track as overlay and all waypoints.
> Okay, you get some shaping points more and all waypoint names start with
> "xx.x km" - see my *Workflow.pdf*.
>
> All files are in this zip: Download.zip
> <https://github.com/devemux86/cruiser/files/10776036/Download.zip>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/devemux86/cruiser/discussions/39#discussioncomment-5022247>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AM5INIOXR47KQ2PVXL2FYQLWYGWXPANCNFSM6AAAAAAU7Z5AFA>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
Willy
***@***.***

**Comment by devemux86** [comment]
Cruiser 3.0.10 with all the discussed improvements will probably be released tomorrow.

You can also test the improved import / export with desktop [Cruiser 3.0.10 (Beta 3)](https://www.dropbox.com/sh/914t158im4b7lxr/AADOeSE1UjaW4FS2PLc7kwAma).
(faster to release Beta on desktop :slightly_smiling_face: )

**Reply by MotoUKRider** [reply]
Ooh, nice.  Hello *Turn Instructions* (which appear instantly) :)
If you've tried to fix the "complete exit" after a route load+clear or map change, that issue still persists here.

**Reply by devemux86** [reply]
> If you've tried to fix the "complete exit" after a route load+clear or map change, that issue still persists here.

I tried, but apparently it still happens?
It is strange as I cannot reproduce it. I will see if something else could be done.

Has anyone else seen any issues exiting the Cruiser GL on Windows?

**Comment by 0709wiwiwi** [comment]
As I do not use openGL (just for a test to help a little) and the fact that
my time is limited, I think anyway that the report by MotorUKdriver is sure
(always) reliable.
On several short test I have noticed too many issues by using Open GL so far.
So I sure keep using the Mapsforge engine for now.
Sorry detailed reports and extra test are not possible currently.
Trust the MotorUkdriver report is the best advice I can give.
Other testers ? Really no one to confirm these issues pse ?

Op ma 20 feb. 2023 om 20:10 schreef Emux ***@***.***>:

> If you've tried to fix the "complete exit" after a route load+clear or map
> change, that issue still persists here.
>
> I tried, but apparently it still happens?
> It is strange as I cannot reproduce it. I will see if something else could
> be done.
>
> Has anyone else seen any issues exiting the Cruiser GL on Windows?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/devemux86/cruiser/discussions/39#discussioncomment-5056658>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AM5INILD243F453A4AFJGDTWYO6SRANCNFSM6AAAAAAU7Z5AFA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Willy
***@***.***

**Comment by devemux86** [comment]
It's recommended (and the default settings):
- Use **VTM (OpenGL)** with **Android** Cruiser
- Use **Mapsforge** with **desktop** Cruiser
(unless you need OpenGL themes like Biker on desktop)

**Comment by devemux86** [comment, marked as answer]
It should work properly with [Cruiser 3.0.10](https://play.google.com/store/apps/details?id=gr.talent.cruiser).

**Reply by MotoUKRider** [reply]
We have a winner :)  Thanks.

**Reply by devemux86** [reply]
Thanks for the report!

**Comment by airald** [comment]
Hi @devemux86 !

routes are again loaded with all waypoint information I put on them. Interface is not the same than v3.0.8 but this settings matches my needs.
![Screenshot_2023-02-24-07-13-39-354_gr talent cruiser](https://user-images.githubusercontent.com/20151719/221349613-51e9f8db-1304-4b1b-9c9e-c3bf58bc585f.jpg)

Thank's a lot!

**Reply by devemux86** [reply]
@airald thanks for the feedback!

> Interface is not the same than v3.0.8 but this settings matches my needs.

The previous interface was a bit complicated for users. :slightly_smiling_face: 

If you need any additional feature, please mention it and we will discuss it.
e.g. if the max number of waypoints should be larger
